clients/go/admin/client.go

```diff
@@ -54,80 +76,78 @@ func GetAdditionalAdminClientConfigOptions(cfg Config) []grpc.DialOption {
-	var tokenURL string
-	tokenURL, err := getTokenEndpointFromAuthServer(ctx, cfg.AuthorizationServerURL)
-	if err != nil || tokenURL == "" {
-		logger.Infof(ctx, "No token URL found from configuration Issuer, looking for token endpoint directly")
+func getAuthenticationDialOption(ctx context.Context, cfg Config, dialOpts []grpc.DialOption) (grpc.DialOption, error) {
+	conn, err := grpc.Dial(cfg.Endpoint.String(), dialOpts...)
+	if err != nil {
+		return nil, err
+	}
+
+	tempClient := NewAuthClient(ctx, conn)
+	tokenURL := cfg.TokenURL
+	if len(tokenURL) == 0 {
+		metadata, err := tempClient.OAuth2Metadata(ctx, &service.OAuth2MetadataRequest{})
 		if err != nil {
-			logger.Errorf(ctx, "Err is %s", err)
-		}
-		tokenURL = cfg.TokenURL
-		if tokenURL == "" {
-			return nil, errors.New("no token endpoint could be found")
+			return nil, fmt.Errorf("failed to fetch auth metadata. Error: %v", err)
 		}
+
+		tokenURL = metadata.TokenEndpoint
+	}
+
+	clientMetadata, err := tempClient.FlyteClient(ctx, &service.FlyteClientRequest{})
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch client metadata. Error: %v", err)
 	}
 
 	secretBytes, err := ioutil.ReadFile(cfg.ClientSecretLocation)
 	if err != nil {
 		logger.Errorf(ctx, "Error reading secret from location %s", cfg.ClientSecretLocation)
 		return nil, err
 	}
+
 	secret := strings.TrimSpace(string(secretBytes))
 
+	scopes := cfg.Scopes
+	if len(scopes) == 0 {
+		scopes = clientMetadata.Scopes
+	}
+
 	ccConfig := clientcredentials.Config{
 		ClientID:     cfg.ClientID,
 		ClientSecret: secret,
 		TokenURL:     tokenURL,
-		Scopes:       cfg.Scopes,
+		Scopes:       scopes,
 	}
 	tSource := ccConfig.TokenSource(ctx)
-	oauthTokenSource := NewCustomHeaderTokenSource(tSource, cfg.AuthorizationHeader)
+	oauthTokenSource := NewCustomHeaderTokenSource(tSource, clientMetadata.AuthorizationMetadataKey)
 	return grpc.WithPerRPCCredentials(oauthTokenSource), nil
 }
 
 func NewAdminConnection(ctx context.Context, cfg Config) (*grpc.ClientConn, error) {
-	var opts []grpc.DialOption
+	opts := make([]grpc.DialOption, 0, 10)
 
 	if cfg.UseInsecureConnection {
 		opts = append(opts, grpc.WithInsecure())
 	} else {
 		// TODO: as of Go 1.11.4, this is not supported on Windows. https://github.com/golang/go/issues/16736
 		creds := credentials.NewClientTLSFromCert(nil, "")
 		opts = append(opts, grpc.WithTransportCredentials(creds))
-		if cfg.UseAuth {
-			logger.Infof(ctx, "Instantiating a token source to authenticate against Admin, ID: %s", cfg.ClientID)
-			jwtDialOption, err := getAuthenticationDialOption(ctx, cfg)
-			if err != nil {
-				return nil, err
-			}
-			opts = append(opts, jwtDialOption)
+	}
+
+	if cfg.UseAuth {
+		logger.Infof(ctx, "Instantiating a token source to authenticate against Admin, ID: %s", cfg.ClientID)
+		jwtDialOption, err := getAuthenticationDialOption(ctx, cfg, append(opts, GetAdditionalAdminClientConfigOptions(cfg)...))
+		if err != nil {
+			return nil, err
```

Comment:
since this common code is now expecting to read Client secret which right now is only applicable for flyte propeller but  wont work for flytectl.
We should probably add this only for client secret authentication and for the 3 legged authentication use the flytectl implementation.
Please see if there is another way to solve this .